plugins/com.google.cloud.tools.eclipse.swtbot/src/com/google/cloud/tools/eclipse/swtbot/SwtBotTreeUtilities.java

```diff
@@ -0,0 +1,305 @@
+      return !isTreeExpanded(tree);
+    }
+  }
+
+  private static class TreeExpandedCondition extends DefaultCondition {
+    private final TreeItem tree;
+
+    private TreeExpandedCondition(TreeItem tree) {
+      this.tree = tree;
+    }
+
+    public String getFailureMessage() {
+      return "Could not expand the tree of " + tree.getText();
+    }
+
+    public boolean test() throws Exception {
+      return isTreeExpanded(tree);
+    }
+  }
+
+  public static SWTBotTreeItem selectTreeItem(SWTBot bot, SWTBotTreeItem tree, String itemText) {
+    waitUntilTreeItemHasItem(bot, tree, itemText);
+    SWTBotTreeItem item = tree.select(itemText);
+    return item;
+  }
+
+  /**
+   * Blocks the caller until all of the direct children of the tree have text. The assumption is
+   * that the tree does not have any "empty" children.
+   * 
+   * TODO: Refactor some of this logic; it follows the same general pattern as
+   * {@link #waitUntilTreeItemHasItem(SWTBot, SWTBotTreeItem, String)}.
+   * 
+   * @param tree the tree to search
+   * @throws TimeoutException if all of the direct children of the tree do not have text within the
+   *         timeout period
+   */
+  public static void waitUntilTreeHasText(SWTBot bot, final SWTBotTreeItem tree)
+      throws TimeoutException {
+    // Attempt #1
+    if (!waitUntilTreeHasTextImpl(bot, tree.widget)) {
+      // Attempt #2: Something went wrong, try to cautiously reopen it.
+      bot.sleep(1000);
+
+      // There isn't a method to collapse, so double-click instead
+      tree.doubleClick();
+      bot.waitUntil(new TreeCollapsedCondition(tree.widget));
+
+      bot.sleep(1000);
+
+      tree.expand();
+      bot.waitUntil(new TreeExpandedCondition(tree.widget));
+
+      if (!waitUntilTreeHasTextImpl(bot, tree.widget)) {
+        printTree(tree.widget);
+        throw new TimeoutException(
+            "Timed out waiting for text of the tree's children, giving up...");
+      }
+    }
+  }
+
+  private static boolean doesTreeItemHaveText(final TreeItem widget) {
+    return UIThreadRunnable.syncExec(new Result<Boolean>() {
+      public Boolean run() {
+        TreeItem[] items = widget.getItems();
+        for (TreeItem item : items) {
+          if (item.getText() == null || item.getText().length() == 0) {
+            return false;
+          }
+        }
+        return true;
+      }
+    });
+  }
+
+  /**
+   * Helper method to check whether a given tree is expanded which can be called from any thread.
+   */
+  private static boolean isTreeExpanded(final TreeItem tree) {
+    return UIThreadRunnable.syncExec(new Result<Boolean>() {
+      public Boolean run() {
+        return tree.getExpanded();
+      }
+    });
+  }
+
+  private static void printTree(final TreeItem tree) {
+    UIThreadRunnable.syncExec(new VoidResult() {
+      public void run() {
+        System.err.println(String.format("%s has %d items:", tree.getText(), tree.getItemCount()));
```

Comment:
I used it in the (broken) m2e wizard tests. It's useful for diagnosing why a test isn't working.